### PR TITLE
Modify 2pt code to use pre-binned catalogs

### PR DESCRIPTION
**Pull request by joezuntz**
This PR depends on #173 and should be considered only after that is merged. 

It:
- splits the `twopoint.py` classes among new files, `twopoint_null_tests.py`, `jackknife.py` and `theory.py`.  The latter two are unchanged.  The original twopoint and the null tests are modified.
- changes the selection code to adds stages using the splitters to make pre-binned catalog files
- changes the twopoint classes to use the pre-binned catalogs
- changes the yaml files to use the new clases

These changes are designed to allow us to use the treecorr low_mem option, and generally reduce memory use in the 2pt stages.  The changes that Judit found for the CosmoDC2 work should all be incorporated.

### CODE REVIEW

File: txpipe/calibrate.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    def run(self):
+
+        #  Extract the configuration parameters
+        cat_type = read_shear_catalog_type(self)
+        use_true = self.config["use_true_shear"]
+        subtract_mean_shear = self.config["subtract_mean_shear"]
+
+        # Prepare the output file, and
```
Comment (chihway) [inline txpipe/calibrate.py:33]: and...?

File: txpipe/calibrate.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        # Prepare the calibrated output catalog
+        f = self.open_output("calibrated_shear_catalog", parallel=True)
+
+        #  we only retain these columns
+        cols = ["ra", "dec", "weight", "g1", "g2"]
+
+        # structure is /shear/bin_1, /shear/bin_2, etc
+        g = f.create_group("shear")
+        g.attrs["nbin"] = nbin
+        g.attrs["nbin_source"] = nbin
```
Comment (chihway) [inline txpipe/calibrate.py:130]: unrelated... what's the difference between nbin and nbin_source?

File: txpipe/calibrate.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                else:
+                    # otherwise just objects in this bin
+                    w = np.where(data["source_bin"] == b)
+                    cal = cals[b]
+
+                # Cut down the data to just this selection for output
+                d = {name: data[name][w] for name in output_cols}
+
+                # Calibrate the shear columns
+                d["g1"], d["g2"] = cal.apply(d["g1"], d["g2"], subtract_mean=subtract_mean_shear)
```
Comment (chihway) [inline txpipe/calibrate.py:106]: Do I understand this correctly that the individual objects will be saved with object-based calibrated shear? a couple things...  1) do we ever want to save the calibration factors for debugging purpose? 2) I thought the way metacal functions, the calibration needs to be estimated via an ensemble of galaxies (either at the 1pt or 2pt level, depending on the statistics) -- is this still consistent with that?

File: txpipe/lens_selector.py [hunk 4]
```diff
@@ -312,9 +313,99 @@ def data_iterator(self):
...
+    """
+    name='TXLensCatalogSplitter'
+
+    inputs = [
+        ('lens_tomography_catalog', TomographyCatalog),
+        ('photometry_catalog', HDFFile),
+    ]
+
+    outputs = [
+        ("calibrated_lens_catalog", HDFFile),
```
Comment (chihway) [inline txpipe/lens_selector.py:329]: is there any calibration for the lenses going on? should we call it something else?

File: txpipe/theory.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
+from .base_stage import PipelineStage
+from .data_types import FiducialCosmology, SACCFile
+import numpy as np
+
+
+class TXTwoPointTheoryReal(PipelineStage):
```
Comment (chihway) [inline txpipe/theory.py:6]: Is it useful to consolidate the two classes in real and fourier? there's a lot of repetition it seems?

Comment (chihway) [reply txpipe/theory.py:6]: nvm, seems different enough it might be too much trouble

File: txpipe/random_cats.py [hunk 6]
```diff
@@ -13,6 +13,7 @@ class TXRandomCat(PipelineStage):
     ]
     outputs = [
         ('random_cats', RandomsCatalog),
+        ('binned_random_cats', RandomsCatalog),
```
Comment (chihway) [inline txpipe/random_cats.py:16]: maybe the calibrated_lens --> binned_lens?

File: txpipe/twopoint.py [hunk 7]
```diff
@@ -705,297 +554,9 @@ def read_metadata(self):
...
-        return s
-
-
-
+    
 
+    
 class TXTwoPointPlots(PipelineStage):
     """
     Make n(z) plots
```
Comment (chihway) [inline txpipe/twopoint.py:562]: this is to make the DV plots right? also wondering if this should be split off to its own function now that theory is split off?

File: txpipe/twopoint.py [hunk 8]
```diff
@@ -385,299 +362,171 @@ def call_treecorr(self, data, meta, i, j, k):
...
-        else:
-            sel = slice(None)
 
-        data['random_ra'] =  group['ra'][sel]
-        data['random_dec'] = group['dec'][sel]
-        data['random_bin'] = group['bin'][sel]
 
-        f.close()
 
     def calculate_area(self, data):
```
Comment (chihway) [inline txpipe/twopoint.py:531]: I know this is not part of the review, but this seems a little dangerous in that 1) nside is fixed 2) for galaxy density, sometimes the weights should be taken into account... maybe at least put a comment for future improvement? also just searching through, is this actually used?

File: txpipe/twopoint_null_tests.py [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+        # self.write_metadata(S, meta)
+
+        # Our data points may currently be in any order depending on which processes
+        # ran which calculations.  Re-order them.
+        S.to_canonical_order()
+
+        # Finally, save the output to Sacc file
+        S.save_fits(self.get_output("gammat_field_center"), overwrite=True)
+
+        # Also make a plot of the data
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:250]: remove?

File: txpipe/twopoint_null_tests.py [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+                dt,
+                ("source2d", "fieldcenter"),
+                dvalue[i],
+                theta=dtheta[i],
+                error=derror[i],
+                npair=dnpair[i],
+                weight=dweight[i],
+            )
+
+        # self.write_metadata(S, meta)
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:241]: remove?

File: txpipe/twopoint_null_tests.py [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+
+        print(S)
+        # Our data points may currently be in any order depending on which processes
+        # ran which calculations.  Re-order them.
+        S.to_canonical_order()
+
+        # Finally, save the output to Sacc file
+        S.save_fits(self.get_output(sacc_file), overwrite=True)
+
+        # Also make a plot of the data
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:438]: remove?

File: txpipe/twopoint_null_tests.py [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+                dvalue[i],
+                theta=dtheta[i],
+                error=derror[i],
+                npair=dnpair[i],
+                weight=dweight[i],
+            )
+
+        self.write_metadata(S, meta)
+
+        print(S)
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:430]: remove?

File: txpipe/twopoint_null_tests.py [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+        # a plot
+        self.write_output_sacc(meta, results)
+        self.write_output_plot(results)
+
+    def write_output_plot(self, results):
+        import matplotlib.pyplot as plt
+
+        d = results[0]
+        dvalue = d.object.xi
+        derror = np.sqrt(d.object.varxi)
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:532]: can we use the full covariance?

File: txpipe/twopoint_null_tests.py [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+                dvalue[i],
+                theta=dtheta[i],
+                error=derror[i],
+                npair=dnpair[i],
+                weight=dweight[i],
+            )
+
+        self.write_metadata(S, meta)
+
+        print(S)
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:606]: remove?

File: txpipe/twopoint_null_tests.py [hunk 15]
```diff
@@ -0,0 +1,614 @@
...
+        # Our data points may currently be in any order depending on which processes
+        # ran which calculations.  Re-order them.
+        S.to_canonical_order()
+
+        # Finally, save the output to Sacc file
+        S.save_fits(self.get_output("gammat_randoms"), overwrite=True)
+
+        # Also make a plot of the data
```
Comment (chihway) [inline txpipe/twopoint_null_tests.py:614]: remove?

File: txpipe/utils/calibrators.py [hunk 16]
```diff
@@ -327,9 +325,6 @@ def apply(self, g1, g2, c1, c2):
 
         c2: array or float
             Shear 2 additive bias component
-
-        subtract_mean: bool
-            whether to subtract the constant c term (default True)
         """
 
         g1 = (g1 / (2 * self.R) - c1)/ (1 + self.K)
```
Comment (chihway) [inline txpipe/utils/calibrators.py:330]: what is K again? isn't R the response already? why is there additional calibration?

**Comment by chihway** [comment]
Ok @joezuntz , I'm through with this PR, added miscellaneous comments, nothing major.

Do let me know if I'm misunderstanding anything!

File: txpipe/calibrate.py (hunk 2 above)
Comment (joezuntz) [reply txpipe/calibrate.py:130]: added a comment

File: txpipe/lens_selector.py (hunk 4 above)
Comment (joezuntz) [reply txpipe/lens_selector.py:329]: Renaming! Also we could change `calibrated_shear_catalog` to `binned_shear_catalog`.  See my conceptual poll on desc-3x2pt!

File: txpipe/utils/calibrators.py (hunk 16 above)
Comment (joezuntz) [reply txpipe/utils/calibrators.py:330]: I've never quite got the difference. Emily may understand it!

File: txpipe/twopoint.py (hunk 8 above)
Comment (joezuntz) [reply txpipe/twopoint.py:531]: Thanks - this isn't used any more so I've removed.  But your questions apply to the place where area is calculated, and I've added a TODO there.

File: txpipe/twopoint.py (hunk 7 above)
Comment (joezuntz) [reply txpipe/twopoint.py:562]: It is - yep - I've fixed the docstring.  Not quite sure which part you mean should be split off?

File: txpipe/twopoint_null_tests.py (hunk 10 above)
Comment (joezuntz) [reply txpipe/twopoint_null_tests.py:241]: uncommented instead.

File: txpipe/twopoint_null_tests.py (hunk 13 above)
Comment (joezuntz) [reply txpipe/twopoint_null_tests.py:532]: Opening a new branch to do this - will do a PR soon.

**Pull request merged**